src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseAssignableTypeForForeachVariableTests.cs

```diff
@@ -0,0 +1,466 @@
+{
+    class Program
+    {   
+        void Main()
+        {
+            foreach (string item in GenerateSequenceAsync())
+            {
+            }
+
+            IEnumerable<IComparable> GenerateSequenceAsync()
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test, GetCSharpResultAt(11, 13).WithArguments("IComparable", "String"));
+        }
+
+        [Fact]
+        public async Task CollectionFromMethodResult_Valid()
+        {
+            var test = @"
+using System;
+using System.Collections.Generic;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            foreach (IComparable item in GenerateSequenceAsync())
+            {
+            }
+
+            IEnumerable<IComparable> GenerateSequenceAsync()
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test);
+        }
+
+        [Fact]
+        public async Task DynamicSameType()
+        {
+            var test = @"
+using System.Collections.Generic;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<dynamic>();
+            foreach (dynamic s in x)
+            {
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test);
+        }
+
+        [Fact]
+        public async Task DynamicToObject()
+        {
+            var test = @"
+using System.Collections.Generic;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<dynamic>();
+            foreach (object s in x)
+            {
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test);
+        }
+
+        [Fact]
+        public async Task DynamicToString()
+        {
+            var test = @"
+using System.Collections.Generic;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<dynamic>();
+            foreach (string s in x)
+            {
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test, GetCSharpResultAt(11, 13).WithArguments("dynamic", "String"));
+        }
+
+        [Fact]
+        public async Task DynamicToVar()
+        {
+            var test = @"
+using System.Collections.Generic;
+
+namespace ConsoleApplication1
+{
+    class Program
+    {   
+        void Main()
+        {
+            var x = new List<dynamic>();
+            foreach (var s in x)
+            {
+            }
+        }
+    }
+}";
+
+            await VerifyCS.VerifyAnalyzerAsync(test);
+        }
+
+        private static DiagnosticResult GetCSharpResultAt(int line, int column)
+#pragma warning disable RS0030 // Do not used banned APIs
```

Comment:
We try to no longer use this syntax but instead either provide a full location with `WithSpan` or by using the markup syntax this helps with having more precise control over how issues are displayed.